db/commitlog/commitlog.cc

```diff
@@ -1946,7 +1952,10 @@ future<> db::commitlog::segment_manager::clear() {
  */
 void db::commitlog::segment_manager::sync() {
     auto f = std::exchange(_background_sync, make_ready_future<>());
-    _background_sync = parallel_for_each(_segments, [](sseg_ptr s) {
+    // #8952 - calls that do sync/cycle can end up altering
+    // _segments (end_flush()->discard_unused())
+    auto def_copy = _segments;
+    _background_sync = parallel_for_each(def_copy, [](sseg_ptr s) {
```

Comment:
nit: how about reusing `sync_all_segments` here?
It adds some debug logging, but it seems appropriate for this path as well.

Reply:
I'd prefer not. At least for me, having the differentiated log helps here, so if so, need to break out into separate subroutine. 
Ok for refactor, but this is bug fix.

Reply:
ok, fair enough. Refactoring can be done as a follow up.